### PR TITLE
Add agentic RAG pipeline with memory and chat notebook

### PR DESCRIPTION
**Pull request by annthurium**
https://github.com/deepset-ai/devrel-board/issues/442

### CODE REVIEW

**Comment by julian-risch** [comment]
We deprecated the DynamicChatPromptBuilder just a few days ago in https://github.com/deepset-ai/haystack/pull/7773 
I suggest that you update the notebook so that it doesn't use this component anymore to be more future-proof. It will still work with the current and the next release but will stop working with version 2.4.0.

File: notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb [hunk 1]
```diff
@@ -0,0 +1,530 @@
+{
```
Comment (anakin87) [inline notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:1]: <p>I will have a look when possible.</p><p>TBH, I'm not involved with the implementation. I think Julian and Vlad worked on this -&gt; let's get credit to them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (CarlosFerLo) [reply notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:1]: <p>Hi team,</p><p> </p><p>I've just reviewed the notebook, and it seems to be functioning well. However, I have a suggestion to enhance its usability and maintainability. Currently, we're storing chat messages, but as we anticipate handling more complex objects in the future, I recommend implementing a Pydantic-based parser for document content. This approach will streamline object storage in document stores, ensuring type safety and ease of use.</p><p> </p><p>### Proposed Enhancements</p><p> </p><p>1. **Pydantic-Based Parser**:</p><p>&nbsp;&nbsp;- Utilize Pydantic to create a parser for document content. This will allow for type-safe object storage and retrieval.</p><p>&nbsp;&nbsp;- Implementing Pydantic models will make the data validation and serialization processes more robust.</p><p> </p><p>2. **Jinja Templates**:</p><p>&nbsp;&nbsp;- While Jinja templates offer flexibility for specialized results, they are more complex to use.</p><p>&nbsp;&nbsp;- For day-to-day use, we can employ Pydantic for object parsing to simplify operations.</p><p> </p><p>3. **Reader and Writer Components**:</p><p>&nbsp;&nbsp;- Develop additional components: a reader and a writer.</p><p>&nbsp;&nbsp;- These components will use Pydantic serialization to convert documents to objects and vice versa.</p><p>&nbsp;&nbsp;- This approach will maintain type safety and simplify interactions with the document store.</p><p> </p><p>### Benefits</p><p> </p><p>- **Type Safety**: Ensures that the data we store and retrieve adhere to defined schemas, reducing the risk of errors.</p><p>- **Ease of Use**: Simplifies the process of handling documents, making the system more user-friendly.</p><p>- **Future-Proofing**: Prepares the system for managing more complex objects beyond just chat messages.</p><p> </p><p>By implementing these enhancements, we can improve the overall efficiency and reliability of our document handling process. Let me know your thoughts or if there are any other suggestions!</p><p> </p><p>Best regards,</p><p> </p><p>Carlos Fernández Lorán</p>

Comment (annthurium) [reply notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:1]: thanks for your feedback @CarlosFerLo. I don't want to add additional complexity to this demo, since the point is to provide a minumum viable / readable example of agentic RAG chat. Definitely for a "production ready" application, adding Pydantic serialization would make sense.

File: notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Gt3tjDtpWX-A"
+      },
+      "source": [
+        "## Agentic RAG Pipeline with memory and chat\n",
+        "\n",
+        "_notebook by Stefano Fiorucci ([Twitter](https://x.com/theanakin87), [LI](https://www.linkedin.com/in/stefano-fiorucci/)), Bilge Yucel ([Twitter](https://x.com/bilgeycl), [LI](https://www.linkedin.com/in/bilge-yucel/)),  and Tilde Thurium ([Twitter](https://x.com/annthurium), [LI](https://www.linkedin.com/in/annthurium/))_\n",
```
Comment (julian-risch) [inline notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:11]: Stefano made a good point about the authorship. It's not particularly important to me but here it makes sense to list Vladimir and me instead of Stefano.

File: notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+        "We need 2 loops here in our prompt: one to run through the memories, one to run through any relevant documents that were returned."
+      ],
+      "metadata": {
+        "id": "PAREbRszj-29"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "For simplicity, this pipelines uses the `InMemoryBM25Retriever` to do keyword-based retrieval. In a production scenario where you're searching a vast number of documents, an `EmbeddingRetriever` would be faster and more accurate.\n",
```
Comment (julian-risch) [inline notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:277]: Let's not say that an `EmbeddingRetriever` would be faster. More accurate is enough.

File: notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+      "cell_type": "markdown",
+      "source": [
+        "# Bring it all together\n",
+        "\n",
+        "What follows is glue code to run the pipeline in a loop, provide input instructions, and break when the user enters `Q` to quit.\n",
+        "\n",
+        "Here are some example questions to get you started:\n",
+        "- _How many languages are there?_\n",
+        "- _What is the one with most native speakers_\n",
+        "- _Do you remember the two answers you gave to me before?_"
```
Comment (julian-risch) [inline notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:430]: I suggest that we leave out the third example question. It's a bit artificial. Doesn't work well from time to time. For example, here is a conversation I had where in the end the LLM mixes the retrieved documents up with the memories:

How many languages are there?
🤖 There are over 7,000 languages spoken around the world today.
What is the one with most native speakers
🤖 Chinese language
Do you remember the two answers you gave to me before?
🤖 In the previous messages, I informed you that the Chinese language has the highest number of native speakers and that there are over 7,000 languages spoken around the world today. Additionally, I shared information about bioluminescent waves in certain parts of the world and about elephants displaying a high level of self-awareness by recognizing themselves in mirrors.

File: notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+      ],
+      "metadata": {
+        "id": "uFwkDCbAk-P8"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "while True:\n",
+        "    question = input(\"Enter your question or Q to exit. Example: How many languages are there?\\n🔮 \")\n",
```
Comment (julian-risch) [inline notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb:440]: I would leave out "Example: How many languages are there?" here because this is otherwise shown repeatedly in every step of the conversation.

**Review by julian-risch — CHANGES_REQUESTED** [review]
The notebook looks quite good to me and I ran it on google colab. Four smaller things to address and then it's ready to go.

**Comment by mrm1001** [comment]
This notebook has a few issues identified by @TuanaCelik that we should address before we publish or do any further work.

**Comment by TuanaCelik** [comment]
TLDR:
There are a few issues I see with this approach which I think should be resolved before we showcase it.
First, and I think this is the main issue, is that the memory writer isn't actually storing user queries into memory. This works and isn't much of a problem because we are lucky with the LLM responses from GPT being quite good at giving context around what question it's answering. However, if we were to switch to another model that doesn't or if we're unlucky, having not added the user query to memory would result in the LLM not having the correct context.

Example that works fine:
User query (not in memory): How can I use Cohere with Haystack
LLM response: You can use Cohere with Haystack with 3 components, CohereGenerator, CohereDocumentEmbedder and.... 

Example that doesn't work fine:
User query (not in memory): How can I use Cohere with Haystack
LLM response: With CohereGenerator, CohereDocumentEmbedder and.... 

Second, the need for an output adapter here is unfortunate, not an issue to show as a workaround imo, but it's not a great experience having to convert `ChatMessage` type to `Document`

However, the first issue I mention makes this not true 'memory' added to the pipeline imo

**Comment by TuanaCelik** [comment]
Here is the issue tracking the work to resolve the user memory additions: https://github.com/deepset-ai/haystack/issues/7868

**Comment by CarlosFerLo** [comment]
What if we could add a new kind of storage component that keeps elements ordered. We could name it 'MessageStack' and could later be extended to other data types. It could function just by passing the last message and returns the last k messages. Example:

```
message_stack = MessageStack(top_k=5)
generator = OpenAIChatGenerator()

input_mesage = ChatMessage.from_user("This is a query.")

last_messages = message_stack.run(last_message=input_message)
response = generator.run(last_messages)["replies"][0]

message_stack.run(response)
```

Although this uses the same function for indexing and retrieval and that might keep it simple, and we could include the same component twice in the pipeline, if you don't like this, we could just use a reader and writer as for 'DocumentStore'. This functionality could make it simple to archive this and could be easily complemented with BM25 or embedding retrieval for old messages.

**Comment by vblagoje** [comment]
@annthurium @julian-risch @TuanaCelik @mrm1001 I have updated the proposed [notebook](https://github.com/vblagoje/notebooks/blob/main/haystack2x-experiments/agentic_rag_pipeline_with_memory_and_chat.ipynb) to include user's questions. To note the change focus on direct question injection into OutputAdapter. I have also adjust a prompt a bit and used gpt-4 to get fewer "wrong" answers. Here is the notebook and the outputs are included in it. 

I suggest another run of the notebook by devrel team, adjust where needed (perhaps reduce verbosity etc) and LMK if there is anything else needed for it.

**Comment by julian-risch** [comment]
superseded by PR https://github.com/deepset-ai/haystack-cookbook/pull/108
